plugins/ag-charts-generate-chart-thumbnail/src/executors/generate/generator/thumbnailGenerator.ts

```diff
@@ -19,59 +20,111 @@ interface Params {
-            const defaultAspectRatio = DEFAULT_THUMBNAIL_WIDTH / DEFAULT_THUMBNAIL_HEIGHT;
-            const aspectRatio = width / height;
-            if (aspectRatio >= defaultAspectRatio) {
-                width = DEFAULT_THUMBNAIL_WIDTH;
-                height = Math.round(width / aspectRatio);
-            } else {
-                height = DEFAULT_THUMBNAIL_HEIGHT;
-                width = Math.round(height * aspectRatio);
+        const { optionsById } = transformPlainEntryFile(entryFile, files['data.js'], theme);
+
+        const jsdom = new JSDOM(`<html><head><style></style></head><body>${files['index.html']}</body></html>`);
+        const { window } = jsdom;
+        const { document } = window;
+
+        const outputCanvas = new Canvas(DEFAULT_THUMBNAIL_WIDTH, DEFAULT_THUMBNAIL_HEIGHT);
+        const outputCtx = outputCanvas.getContext('2d');
+
+        const chartContainers = Array.from(document.querySelectorAll<HTMLElement>('[id]'));
+
+        let rows = 1;
+        let columns = 1;
+        chartContainers.forEach((container) => {
+            const gridArea = getGridArea(container);
+            if (gridArea == null) return;
+            rows = Math.max(gridArea.row + 1, rows);
+            columns = Math.max(gridArea.column + 1, columns);
+        });
+
+        const promises = chartContainers.map(async (container) => {
+            const gridArea = getGridArea(container);
+            const row = gridArea?.row ?? 0;
+            const column = gridArea?.column ?? 0;
+
+            const width = (DEFAULT_THUMBNAIL_WIDTH / columns) | 0;
+            const height = (DEFAULT_THUMBNAIL_HEIGHT / rows) | 0;
+            const x0 = (width * column) | 0;
+            const y0 = (width * row) | 0;
+
+            const mockCtx = mockCanvas.setup({
+                width,
+                height,
+                document,
+                window: window as any,
+                mockText: false,
+            });
+
+            const chartOptions = {
+                animation: { enabled: false },
+                document,
+                window,
+                width,
+                height,
+            } as any;
+
+            const chartProxy = AgCharts.create(chartOptions);
+
+            let options = optionsById.get(container.id);
+            if (options == null) {
+                throw new Error(`No options found for container with id "${container.id}"`);
+            }
+
+            options = {
+                ...options,
+                theme,
+                animation: { enabled: false },
+                document,
+                window,
+                width,
+                height,
+            } as any;
+
+            AgCharts.update(chartProxy, options);
+            const chart = (chartProxy as any).chart;
+            await chart.waitForUpdate(5_000);
+
+            // BGRA
+            const imageData = mockCtx.ctx.nodeCanvas.toBuffer('raw');
```

Comment:
If we could get the context from the mockCtx, we could use `getImageData` to speed this up

Reply:
Should be possible trivially?